File: tests/ecmascript/test-bug-object-binding-proxy.js
/*
 *  Object binding (global object or with statement) can be a Proxy object
 *  but Duktape 1.2.x doesn't handle that correctly.  This test demonstrates
 *  the bug for with-statement, C code is needed to demonstrate the global
 *  object issue.  See: https://github.com/svaarala/duktape/issues/221.
 */

/*===
in with (obj) statement
123
bar
has print
in with (proxy) statement
has print
has dummyGlobal
123
has print
has foo
get foo
bar
===*/

var dummyGlobal = 123;

function test() {
    var obj = { foo: 'bar' };
    var proxy = new Proxy(obj, {
        get: function get(targ, key, recv) {
            print('get ' + key);
            return targ[key];
        },
        has: function has(targ, key) {
            print('has ' + key);
            return key in targ;
        }
    });

    with (obj) {
        print('in with (obj) statement');
        print(dummyGlobal);
        print(foo);
    }

    with (proxy) {
        print('in with (proxy) statement');
        print(dummyGlobal);
        print(foo);
    }
}

try {
    test();
} catch (e) {
    print(e.stack || e);
}

File: tests/api/test-bug-object-binding-proxy.c
/*
 *  Counterpart to test-bug-object-binding-proxy.js: demonstrate bug in
 *  Duktape 1.2.x where a Proxy doesn't work as a global object replacement
 *  as expected.  See: https://github.com/svaarala/duktape/issues/221.
 */

/*===
*** test_1 (duk_safe_call)
get print
hello using this.print
rc=0: undefined
has print
get print
hello using print
rc=0: undefined
final top: 0
==> rc=0, result='undefined'
===*/

static duk_ret_t test_1(duk_context *ctx) {
	duk_int_t rc;

	duk_eval_string(ctx,
		"new Proxy(new Function('return this')() /* global object */, {\n"
		"    get: function (targ, key, recv) {\n"
		"        targ.print('get ' + key); /* careful to call print() directly without invoking Proxy! */\n"
		"        return targ[key];\n"
		"    },\n"
		"    has: function (targ, key) {\n"
		"        targ.print('has ' + key);\n"
		"        return key in targ;\n"
		"    }\n"
		"});\n"
	);
	duk_set_global_object(ctx);

	/* This worked in Duktape 1.2.x because this.print goes through a
	 * property lookup with proper Proxy support.
	 */
	rc = duk_peval_string(ctx, "this.print('hello using this.print');");
	printf("rc=%ld: %s\n", (long) rc, duk_safe_to_string(ctx, -1));
	duk_pop(ctx);

	/* This didn't work because print is looked up using a slow path
	 * GETVAR which goes through variable lookup code which didn't
	 * support Proxy properly.
	 */
	rc = duk_peval_string(ctx, "print('hello using print');");
	printf("rc=%ld: %s\n", (long) rc, duk_safe_to_string(ctx, -1));
	duk_pop(ctx);

	printf("final top: %ld\n", (long) duk_get_top(ctx));
	return 0;
}

void test(duk_context *ctx) {
	TEST_SAFE_CALL(test_1);
}
